src/RichText.js

```diff
@@ -240,3 +275,50 @@ export function textOffsetsToSelectionState({start, end}: {start: number, end: n
 
     return selectionState;
 }
+
+// modified version of https://github.com/draft-js-plugins/draft-js-plugins/blob/master/draft-js-emoji-plugin/src/modifiers/attachImmutableEntitiesToEmojis.js
+export function attachImmutableEntitiesToEmoji(editorState: EditorState): EditorState {
+    const contentState = editorState.getCurrentContent();
+    const blocks = contentState.getBlockMap();
+    let newContentState = contentState;
+
+    blocks.forEach((block) => {
+        const plainText = block.getText();
+
+        const addEntityToEmoji = (start, end) => {
+            const existingEntityKey = block.getEntityAt(start);
+            if (existingEntityKey) {
+                // avoid manipulation in case the emoji already has an entity
+                const entity = Entity.get(existingEntityKey);
+                if (entity && entity.get('type') === 'emoji') {
+                    return;
+                }
+            }
+
+            const selection = SelectionState.createEmpty(block.getKey())
+                .set('anchorOffset', start)
+                .set('focusOffset', end);
+            const emojiText = plainText.substring(start, end);
+            const entityKey = Entity.create('emoji', 'IMMUTABLE', { emojiUnicode: emojiText });
+            newContentState = Modifier.replaceText(
+                newContentState,
+                selection,
+                emojiText,
+                null,
+                entityKey,
```

Comment:
:| For consistency:
http://babeljs.io/docs/plugins/syntax-trailing-function-commas/

On Fri, Jul 8, 2016 at 7:48 PM David Baker <notifications@github.com> wrote:

> In src/RichText.js
> <https://github.com/matrix-org/matrix-react-sdk/pull/335#discussion_r70079753>
> :
>
> > +                if (entity && entity.get('type') === 'emoji') {
> > +                    return;
> > +                }
> > +            }
> > +
> > +            const selection = SelectionState.createEmpty(block.getKey())
> > +                .set('anchorOffset', start)
> > +                .set('focusOffset', end);
> > +            const emojiText = plainText.substring(start, end);
> > +            const entityKey = Entity.create('emoji', 'IMMUTABLE', { emojiUnicode: emojiText });
> > +            newContentState = Modifier.replaceText(
> > +                newContentState,
> > +                selection,
> > +                emojiText,
> > +                null,
> > +                entityKey,
>
> Confusingly, trailing commas are now valid ES6 in object literals, array
> literals and imports, but not function calls...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/matrix-org/matrix-react-sdk/pull/335/files/b33452216875ba3ffd09747158bd5e1a12512e96#r70079753>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAOlRLQEKE_mgIb9QfvZaMrjpUK4ZyUFks5qTlqsgaJpZM4JHyQd>
> .
>